build failure on macOS

fresh clone of master branch at 19f81ac
```
[ 75%] Building CXX object lib/ExecutionEngine/Orc/CMakeFiles/LLVMOrcJIT.dir/OrcMCJITReplacement.cpp.o
[ 75%] Building CXX object lib/ExecutionEngine/Orc/CMakeFiles/LLVMOrcJIT.dir/OrcRemoteTargetRPCAPI.cpp.o
In file included from /Users/sachs/src/julia-master/deps/srccache/llvm-3.9.1/lib/ExecutionEngine/Orc/OrcRemoteTargetRPCAPI.cpp:10:
In file included from /Users/sachs/src/julia-master/deps/srccache/llvm-3.9.1/include/llvm/ExecutionEngine/Orc/OrcRemoteTargetRPCAPI.h:21:
In file included from /Users/sachs/src/julia-master/deps/srccache/llvm-3.9.1/include/llvm/ExecutionEngine/Orc/RPCUtils.h:35:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/future:1447:23: error:
      'future_error' is unavailable: introduced in macOS 10.8
                      future_error(make_error_code(future_errc::broken_promise))
                      ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/future:502:63: note:
      'future_error' has been explicitly marked unavailable here
class _LIBCPP_EXCEPTION_ABI _LIBCPP_AVAILABILITY_FUTURE_ERROR future_error
                                                              ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/future:1621:23: error:
      'future_error' is unavailable: introduced in macOS 10.8
                      future_error(make_error_code(future_errc::broken_promise))
                      ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/future:502:63: note:
      'future_error' has been explicitly marked unavailable here
class _LIBCPP_EXCEPTION_ABI _LIBCPP_AVAILABILITY_FUTURE_ERROR future_error
                                                              ^
2 errors generated.
make[4]: *** [lib/ExecutionEngine/Orc/CMakeFiles/LLVMOrcJIT.dir/OrcRemoteTargetRPCAPI.cpp.o] Error 1
make[3]: *** [lib/ExecutionEngine/Orc/CMakeFiles/LLVMOrcJIT.dir/all] Error 2
make[2]: *** [all] Error 2
make[1]: *** [scratch/llvm-3.9.1/build_Release/build-compiled] Error 2
make: *** [julia-deps] Error 2
```

--- a/DISTRIBUTING.md
+++ b/DISTRIBUTING.md
@@ -104,9 +104,10 @@ completely self-contained Julia.app.
 
 Note that if you want your `.app` to be able to run on OSX 10.6 Snow
 Leopard, you must pass `USE_SYSTEM_LIBUNWIND=1` as one of the make
-variables passed to both `make` processes.  This disables the use of
+variables passed to both `make` processes. This disables the use of
 `libosxunwind`, a more modern libunwind that relies on OS features
-available only in 10.7+.  This is the reason why we support 10.7+ [while we did support 10.6 prior to Julia 0.3.0](http://julialang.org/downloads/platform.html).
+available only in 10.7+. Furthermore, support for OSX 10.6 and 10.7
+requires that Julia is built with `USE_LIBCPP=0`.
 
 Windows
 -------

--- a/Make.inc
+++ b/Make.inc
@@ -403,9 +403,9 @@ DEBUGFLAGS := -O0 -g -DJL_DEBUG_BUILD -fstack-protector-all
 SHIPFLAGS := -O3 -g
 ifeq ($(OS), Darwin)
 ifeq ($(USE_LIBCPP), 1)
-CC += -stdlib=libc++ -mmacosx-version-min=10.7
-CXX += -stdlib=libc++ -mmacosx-version-min=10.7
-FC += -mmacosx-version-min=10.7
+CC += -stdlib=libc++ -mmacosx-version-min=10.8
+CXX += -stdlib=libc++ -mmacosx-version-min=10.8
+FC += -mmacosx-version-min=10.8
 else
 CC += $(STDLIBCPP_FLAG) -mmacosx-version-min=10.6
 CXX += $(STDLIBCPP_FLAG) -mmacosx-version-min=10.6

--- a/Make.inc
+++ b/Make.inc
@@ -403,12 +403,14 @@ DEBUGFLAGS := -O0 -g -DJL_DEBUG_BUILD -fstack-protector-all
 SHIPFLAGS := -O3 -g
 ifeq ($(OS), Darwin)
 ifeq ($(USE_LIBCPP), 1)
-CC += -stdlib=libc++ -mmacosx-version-min=10.8
-CXX += -stdlib=libc++ -mmacosx-version-min=10.8
-FC += -mmacosx-version-min=10.8
+MACOSX_VERSION_MIN := 10.8
+CC += -stdlib=libc++ -mmacosx-version-min=$(MACOSX_VERSION_MIN)
+CXX += -stdlib=libc++ -mmacosx-version-min=$(MACOSX_VERSION_MIN)
+FC += -mmacosx-version-min=$(MACOSX_VERSION_MIN)
 else
-CC += $(STDLIBCPP_FLAG) -mmacosx-version-min=10.6
-CXX += $(STDLIBCPP_FLAG) -mmacosx-version-min=10.6
+MACOSX_VERSION_MIN := 10.6
+CC += $(STDLIBCPP_FLAG) -mmacosx-version-min=$(MACOSX_VERSION_MIN)
+CXX += $(STDLIBCPP_FLAG) -mmacosx-version-min=$(MACOSX_VERSION_MIN)
 endif
 JCPPFLAGS += -D_LARGEFILE_SOURCE -D_DARWIN_USE_64_BIT_INODE=1
 endif

